File: app/src/main/kotlin/jb/openware/app/tools/LinkSpan.java
package in.afi.codekosh.tools;

import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.RippleDrawable;
import android.graphics.drawable.ShapeDrawable;
import android.graphics.drawable.shapes.RoundRectShape;
import android.util.AttributeSet;
import android.util.StateSet;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import in.afi.codekosh.R;

public class LinkSpan extends LinearLayout {
    // Variables
    public static float density = 1;
    private TextView textView1;
    private TextView textView2;

    public LinkSpan(Context context) {
        super(context);
        initializeViews(context);
    }

    public LinkSpan(Context context, AttributeSet attrs) {
        super(context, attrs);
        initializeViews(context);
    }

    public LinkSpan(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initializeViews(context);
    }

    public static int dp(float value) {
        if (value == 0) {
            return 0;
        }
        return (int) Math.ceil(density * value);
    }

    public static Drawable getRoundRectSelectorDrawable(int color) {
        Drawable maskDrawable = createRoundRectDrawable(dp((15)), 0xffffffff);
        ColorStateList colorStateList = new ColorStateList(new int[][]{StateSet.WILD_CARD}, new int[]{(color & 0x00ffffff) | 0x19000000});
        return new RippleDrawable(colorStateList, null, maskDrawable);
    }

    public static Drawable createRoundRectDrawable(int rad, int defaultColor) {
        ShapeDrawable defaultDrawable = new ShapeDrawable(new RoundRectShape(new float[]{rad, rad, rad, rad, rad, rad, rad, rad}, null, null));
        defaultDrawable.getPaint().setColor(defaultColor);
        return defaultDrawable;
    }

    private void initializeViews(Context context) {
        View.inflate(context, R.layout.link_span_layout, this);
        textView1 = findViewById(R.id.textView1);
        textView2 = findViewById(R.id.textView2);
        int getColor = 0xff4991cc;
        textView2.setBackground(getRoundRectSelectorDrawable(getColor));
        int padding = (int) getResources().getDimension(R.dimen.link_span_padding);
        setPaddingRelative(padding, padding, padding, padding);
    }

    public void setFirstText(String text) {
        textView1.setText(text);
    }

    public void setSecondText(String text) {
        textView2.setText(text);
    }

    public void setFirstTextColor(int color) {
        textView1.setTextColor(color);
    }

    public void setSecondTextColor(int color) {
        textView2.setTextColor(color);
    }

    public void setTextSize(float size) {
        textView1.setTextSize(size);
        textView2.setTextSize(size);
    }

    public void setOnClickListener(OnClickListener listener) {
        textView2.setOnClickListener(listener);
    }
}

File: app/src/main/kotlin/jb/openware/app/ui/components/LinkSpan.kt
package jb.openware.app.ui.components

import android.content.Context
import android.content.res.ColorStateList
import android.graphics.drawable.Drawable
import android.graphics.drawable.RippleDrawable
import android.graphics.drawable.ShapeDrawable
import android.graphics.drawable.shapes.RoundRectShape
import android.util.AttributeSet
import android.util.StateSet
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import jb.openware.app.R
import kotlin.math.ceil

class LinkSpan @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : LinearLayout(context, attrs, defStyleAttr) {

    private val textView1: TextView
    private val textView2: TextView

    init {
        inflate(context, R.layout.link_span_layout, this)
        textView1 = findViewById(R.id.textView1)
        textView2 = findViewById(R.id.textView2)

        val color = 0xFF4991CC.toInt()
        textView2.background = getRoundRectSelectorDrawable(color)

        val padding = resources.getDimensionPixelSize(R.dimen.link_span_padding)
        setPaddingRelative(padding, padding, padding, padding)
    }

    fun setFirstText(text: String) {
        textView1.text = text
    }

    fun setSecondText(text: String) {
        textView2.text = text
    }

    fun setFirstTextColor(color: Int) {
        textView1.setTextColor(color)
    }

    fun setSecondTextColor(color: Int) {
        textView2.setTextColor(color)
    }

    fun setTextSize(sizeSp: Float) {
        textView1.textSize = sizeSp
        textView2.textSize = sizeSp
    }

    override fun setOnClickListener(l: OnClickListener?) {
        // Delegate clicks only to the second text, same as your Java behavior
        textView2.setOnClickListener(l)
    }

    companion object {
        @JvmField
        var density: Float = 1f

        @JvmStatic
        fun dp(value: Float): Int {
            if (value == 0f) return 0
            return ceil(density * value).toInt()
        }

        @JvmStatic
        fun getRoundRectSelectorDrawable(color: Int): Drawable {
            val radius = dp(15f)
            val maskDrawable = createRoundRectDrawable(radius, 0xFFFFFFFF.toInt())
            val colorStateList = ColorStateList(
                arrayOf(StateSet.WILD_CARD),
                intArrayOf((color and 0x00FFFFFF) or 0x19000000)
            )
            return RippleDrawable(colorStateList, null, maskDrawable)
        }

        @JvmStatic
        fun createRoundRectDrawable(rad: Int, defaultColor: Int): Drawable {
            val r = rad.toFloat()
            val radii = floatArrayOf(r, r, r, r, r, r, r, r)
            return ShapeDrawable(RoundRectShape(radii, null, null)).apply {
                paint.color = defaultColor
            }
        }
    }
}
